src/coprocessor/dag/batch/executors/selection_executor.rs

```diff
@@ -0,0 +1,639 @@
+                },
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::from(vec![
+                        VectorValue::Int(vec![None]),
+                        VectorValue::Real(vec![None]),
+                    ]),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(true),
+                },
+            ],
+        )
+    }
+
+    /// Tests the scenario that there is no predicate or there is a predicate but always returns
+    /// true (no data is filtered).
+    #[test]
+    fn test_no_predicate_or_predicate_always_true() {
+        // Build a selection executor without predicate.
+        let exec_no_predicate =
+            |src_exec: MockExecutor| BatchSelectionExecutor::new_for_test(src_exec, vec![]);
+
+        // Build a selection executor with a predicate that always returns true.
+        let exec_predicate_true = |src_exec: MockExecutor| {
+            let predicate = RpnExpressionBuilder::new()
+                .push_constant(1i64, FieldTypeTp::LongLong)
+                .build();
+            BatchSelectionExecutor::new_for_test(src_exec, vec![predicate])
+        };
+
+        let executor_builders: Vec<Box<dyn std::boxed::FnBox(MockExecutor) -> _>> =
+            vec![Box::new(exec_no_predicate), Box::new(exec_predicate_true)];
+
+        for exec_builder in executor_builders {
+            let src_exec = make_src_executor_using_fixture_1();
+
+            let mut exec = exec_builder(src_exec);
+
+            // The selection executor should return data as it is.
+
+            let r = exec.next_batch(1);
+            assert_eq!(r.data.rows_len(), 2);
+            assert_eq!(r.data.columns_len(), 2);
+            assert_eq!(r.data[0].decoded().as_int_slice(), &[Some(1), None]);
+            assert_eq!(r.data[1].decoded().as_real_slice(), &[None, Some(7.0)]);
+            assert!(!r.is_drained.unwrap());
+
+            let r = exec.next_batch(1);
+            assert_eq!(r.data.rows_len(), 0);
+            assert!(!r.is_drained.unwrap());
+
+            let r = exec.next_batch(1);
+            assert_eq!(r.data.rows_len(), 1);
+            assert_eq!(r.data.columns_len(), 2);
+            assert_eq!(r.data[0].decoded().as_int_slice(), &[None]);
+            assert_eq!(r.data[1].decoded().as_real_slice(), &[None]);
+            assert!(r.is_drained.unwrap());
+        }
+    }
+
+    /// Tests the scenario that the predicate always returns false.
+    #[test]
+    fn test_predicate_always_false() {
+        let src_exec = make_src_executor_using_fixture_1();
+
+        let predicate = RpnExpressionBuilder::new()
+            .push_constant(0i64, FieldTypeTp::LongLong)
+            .build();
+        let mut exec = BatchSelectionExecutor::new_for_test(src_exec, vec![predicate]);
+
+        // The selection executor should always return empty rows.
+
+        let r = exec.next_batch(1);
+        assert_eq!(r.data.rows_len(), 0);
```

Comment:
Could we use a loop to avoid the repeated code?

Reply:
it only repeats for the first 2 call. The third call is different (`is_drained == true`).